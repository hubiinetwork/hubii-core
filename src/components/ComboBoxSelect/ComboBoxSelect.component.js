/**
*
* ComboBoxSelect
*
*/

import React from 'react';
import PropTypes from 'prop-types';
import { StyledSelect } from './ComboBoxSelect.style';
import { Option } from '../ui/Select';

class ComboBoxSelect extends React.PureComponent { // eslint-disable-line react/prefer-stateless-function
  constructor(props) {
    super(props);
    this.onInputKeyDown = this.onInputKeyDown.bind(this);
    this.handleSelect = this.handleSelect.bind(this);
    this.onBlur = this.onBlur.bind(this);
    this.handleFilterOption = this.handleFilterOption.bind(this);
    this.state = {
      options: [],
      selectValue: this.props.options[0] ? this.props.options[0].name : '',
      notFoundMessage: null,
    };
  }

  componentWillReceiveProps(nextProps) {
    this.setState({
      options: nextProps.options,
    });
  }

  onBlur() {
    const { selectValue } = this.state;
    const { options } = this.props;
    const value = selectValue.trim();
    const existingContact = options.find((option) => option.name.toLowerCase() === value.toLowerCase() || option.value === selectValue);
    if (!this.props.addInputValidator(value) && !existingContact) {
      this.setState({ selectValue: this.props.invalidAdditionMessage });
    } else if (existingContact) {
      this.setState({ selectValue: existingContact.name });
    } else {
      this.setState({
        selectValue: value,
      });
    }
    this.props.handleSelect(value);
  }

  onInputKeyDown(event) {
    const { options } = this.props;
    const value = event.target.value.trim();
    if (event.key === 'Enter' && !this.props.addInputValidator(value) && options.every((option) => option.name.toLowerCase() !== value.toLowerCase())) {
      this.setState({ notFoundMessage: this.props.invalidAdditionMessage, selectValue: options[0] ? options[0].name : '' });
    } else if (event.key === 'Enter' && options.every((option) => option.value !== value && option.name.toLowerCase() !== value)) {
      this.setState((prevState) => ({
        options: [...prevState.options, { name: value, value }],
      }));
    }
  }

  handleSelect(value) {
    this.setState({
      selectValue: value,
    });
    this.props.handleSelect(value);
  }

  handleFilterOption(input, option) {
    return option.props.children.toLowerCase().indexOf(input.toLowerCase().trim()) >= 0 || option.props.value === input.trim();
  }

  render() {
    const { disabled } = this.props;
    const { notFoundMessage, options } = this.state;
    return (
      <StyledSelect
        disabled={disabled}
        showSearch
        recipient
        onInputKeyDown={this.onInputKeyDown}
        onSearch={(value) => this.setState({ selectValue: value, notFoundMessage: value })}
        onSelect={this.handleSelect}
        notFoundContent={notFoundMessage}
        onBlur={this.onBlur}
        filterOption={(input, option) => this.handleFilterOption(input, option)}
        value={this.state.selectValue}
      >
        {options.map((option) => (
          <Option key={option.name} value={option.value}>
            {option.name}
          </Option>
        ))}
      </StyledSelect>
    );
  }
}

ComboBoxSelect.propTypes = {
  /**
   * Determines if the ComboBoxSelect is disabled
   */
  disabled: PropTypes.bool.isRequired,
  /**
   * Options are an array of object, where there must be a "name" and "value" field
   */
  options: PropTypes.array,
  /**
   * handles events onSelect
   */
  handleSelect: PropTypes.func.isRequired,
  /**
   * Input into select input validator
   */
  addInputValidator: PropTypes.func.isRequired,
  /**
   * Message if there is an invalid addition
   */
  invalidAdditionMessage: PropTypes.string.isRequired,
};

export default ComboBoxSelect;
