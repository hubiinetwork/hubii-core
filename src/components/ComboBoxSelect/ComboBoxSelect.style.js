import styled from 'styled-components';
import Select from 'components/ui/Select';

export const StyledSelect = styled(Select)`
  .ant-select-selection{
    box-shadow: none;
  }
  color: ${({ theme }) => theme.palette.info};
  .ant-select-selection-selected-value {
    color: ${({ theme }) => theme.palette.light};
  }
`;
