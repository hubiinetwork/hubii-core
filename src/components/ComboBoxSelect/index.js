/**
*
* ComboBoxSelect
*
*/

import React from 'react';
import PropTypes from 'prop-types';
import { isAddressMatch } from 'utils/wallet';
import { Option } from 'components/ui/Select';
import { StyledSelect } from './style';

class ComboBoxSelect extends React.PureComponent { // eslint-disable-line react/prefer-stateless-function
  constructor(props) {
    super(props);
    this.state = {
      options: [],
      selectValue: '',
      notFoundMessage: null,
    };
    this.onInputKeyDown = this.onInputKeyDown.bind(this);
    this.handleSelect = this.handleSelect.bind(this);
    this.onBlur = this.onBlur.bind(this);
    this.handleFilterOption = this.handleFilterOption.bind(this);
  }

  componentWillMount() {
    this.setState({
      options: this.props.options,
    });
  }

  componentDidUpdate(prevProps) {
    if (this.props.options.length !== prevProps.options.length) {
      this.setState({ // eslint-disable-line react/no-did-update-set-state
        options: this.props.options,
      });
    }
  }

  // When the component is onBlurr'ed this function will be called. This function is important
  // as without it, if the user just clicks out without pressing "Enter" nothing will happen.
  onBlur() {
    const { selectValue } = this.state;
    const { options } = this.props;
    const value = selectValue.trim();
    // check if there is an exisiting contact for the input
    const existingContact = options.find((option) => option.name.toLowerCase() === value.toLowerCase() || isAddressMatch(option.value, selectValue));
    // set the input value to be a Invalid Addition Message as the input is not a valid input and nor is it an exisiting option
    if (!this.props.addInputValidator(value) && !existingContact) {
      this.setState({ selectValue: this.props.invalidAdditionMessage });
    } else if (existingContact) {
      this.setState({ selectValue: existingContact.name });
    } else {
    // this is for if there is a valid input but there is no existing option
      this.setState({
        selectValue: value,
      });
    }
    this.props.handleSelect(value);
  }

  onInputKeyDown(event) {
    const { options } = this.props;
    const value = event.target.value.trim();
    // basic check to see if it is an invalid input with no exisiting option
    if (event.key === 'Enter' && !this.props.addInputValidator(value) && options.every((option) => option.name.toLowerCase() !== value.toLowerCase())) {
      this.setState({ notFoundMessage: this.props.invalidAdditionMessage, selectValue: options[0] ? options[0].name : '' });
    } else if (event.key === 'Enter' && options.every((option) => !isAddressMatch(option.value, value) && option.name.toLowerCase() !== value)) {
      // will only this option if it is a valid input and it is not an existing option
      this.setState((prevState) => ({
        options: [...prevState.options, { name: value, value }],
      }));
    }
  }

  handleSelect(value) {
    this.setState({
      selectValue: value,
    });
    this.props.handleSelect(value);
  }

  handleFilterOption(input, option) {
    return option.props.children.toLowerCase().indexOf(input.toLowerCase().trim()) >= 0 || isAddressMatch(option.props.value, input.trim());
  }

  render() {
    const { disabled } = this.props;
    const { notFoundMessage, options } = this.state;
    return (
      <StyledSelect
        disabled={disabled}
        showSearch
        recipient
        onInputKeyDown={this.onInputKeyDown}
        onSearch={(value) => this.setState({ selectValue: value, notFoundMessage: value })}
        onSelect={this.handleSelect}
        notFoundContent={notFoundMessage}
        onBlur={this.onBlur}
        filterOption={(input, option) => this.handleFilterOption(input, option)}
        value={this.state.selectValue}
      >
        {options.map((option) => (
          <Option key={option.name} value={option.value}>
            {option.name}
          </Option>
        ))}
      </StyledSelect>
    );
  }
}

ComboBoxSelect.propTypes = {
  disabled: PropTypes.bool.isRequired,
  options: PropTypes.array,
  handleSelect: PropTypes.func.isRequired,
  addInputValidator: PropTypes.func.isRequired,
  invalidAdditionMessage: PropTypes.string.isRequired,
};

export default ComboBoxSelect;
