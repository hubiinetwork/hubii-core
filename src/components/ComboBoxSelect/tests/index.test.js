import React from 'react';
import { shallow } from 'enzyme';

import ComboBoxSelect from '../index';

describe('<ComboBoxSelect />', () => {
  const params = {
    disabled: false,
    options: [
      {
        name: 'mike',
        value: '0x324234',
      },
      {
        name: 'jake',
        value: '0x444224',
      },
    ],
    handleSelect: jest.fn(),
    addInputValidator: jest.fn(),
    invalidAdditionMessage: 'message',
  };
  let wrapper;
  let instance;
  beforeEach(() => {
    wrapper = shallow(<ComboBoxSelect {...params} />);
    instance = wrapper.instance();
    instance.setState({
      options: [
        {
          name: 'mike',
          value: '0x324234',
        },
        {
          name: 'jake',
          value: '0x444224',
        },
      ],
      selectValue: 'mike',
      notFoundMessage: 'Not Found',
    });
  });
  afterEach(() => {
    jest.clearAllMocks();
  });
  it('should correctly render', () => {
    expect(wrapper).toMatchSnapshot();
  });
  describe('#componentWillReceiveProps', () => {
    it('should update the options if the props have changed', () => {
      const newOptions = [
        {
          name: 'mike',
          value: '0x324234',
        },
      ];
      instance.componentWillReceiveProps({ options: newOptions });
      expect(instance.state.options).toEqual(newOptions);
    });
  });
  describe('#onBlur', () => {
    it('should provide an error message if there is an invalid input', () => {
      instance.setState({
        selectValue: 'no one',
      });
      instance.props.addInputValidator.mockReturnValueOnce(false);
      instance.onBlur();
      expect(instance.state.selectValue).toEqual(instance.props.invalidAdditionMessage);
      expect(instance.props.handleSelect).toHaveBeenCalledTimes(1);
    });
    it('should display the relevant option name in the input if the associated value is used', () => {
      instance.setState({
        selectValue: '0x324234',
      });
      const expectedSelectValue = 'mike';
      instance.props.addInputValidator.mockReturnValueOnce(true);
      instance.onBlur();
      expect(instance.state.selectValue).toEqual(expectedSelectValue);
      expect(instance.props.handleSelect).toHaveBeenCalledTimes(1);
    });
    it('should display the selected value if it is a valid input but not an option', () => {
      instance.setState({
        selectValue: '0x324sadasd234',
      });
      const expectedSelectValue = '0x324sadasd234';
      instance.props.addInputValidator.mockReturnValueOnce(true);
      instance.onBlur();
      expect(instance.state.selectValue).toEqual(expectedSelectValue);
      expect(instance.props.handleSelect).toHaveBeenCalledTimes(1);
    });
    it('should display the relevant name, if the input is the relevant name option', () => {
      instance.setState({
        selectValue: 'Mike',
      });
      const expectedSelectValue = 'mike';
      instance.props.addInputValidator.mockReturnValueOnce(false);
      instance.onBlur();
      expect(instance.state.selectValue).toEqual(expectedSelectValue);
      expect(instance.props.handleSelect).toHaveBeenCalledTimes(1);
    });
  });
  describe('#handleSelect', () => {
    it('should set the selectValue state, and call the this.props.handleSelect function', () => {
      const value = 'mike';
      instance.handleSelect(value);
      expect(instance.state.selectValue).toEqual(value);
      expect(instance.props.handleSelect).toHaveBeenCalledTimes(1);
    });
  });
  describe('#onInputKeyDown', () => {
    it('should provide invalid message if key inputted is \'Enter\' and the input is invalid', () => {
      const event = {
        target: {
          value: 'hello',
        },
        key: 'Enter',
      };
      instance.props.addInputValidator.mockReturnValueOnce(false);
      instance.onInputKeyDown(event);
      expect(instance.state.notFoundMessage).toEqual(instance.props.invalidAdditionMessage);
    });
    it('should add option if when \'Enter\' the input is valid', () => {
      const event = {
        target: {
          value: '0x123213',
        },
        key: 'Enter',
      };
      const newOptions = [
        {
          name: 'mike',
          value: '0x324234',
        },
        {
          name: 'jake',
          value: '0x444224',
        },
        {
          name: '0x123213',
          value: '0x123213',
        },
      ];
      instance.props.addInputValidator.mockReturnValueOnce(true);
      instance.onInputKeyDown(event);
      expect(instance.state.options).toEqual(newOptions);
    });
  });
  describe('#handleFilterOption', () => {
    const option = {
      props: {
        children: 'mike',
        value: '0x32324',
      },
    };
    it('should return true if the associated option name matches the input', () => {
      const input = 'mike';
      const expectedResult = true;
      expect(instance.handleFilterOption(input, option)).toEqual(expectedResult);
    });
    it('should return true if the associated option value matches the input', () => {
      const input = '0x32324';
      const expectedResult = true;
      expect(instance.handleFilterOption(input, option)).toEqual(expectedResult);
    });
    it('should return false if the associated option name or value does not match the input', () => {
      const input = '0xasdasd32324';
      const expectedResult = false;
      expect(instance.handleFilterOption(input, option)).toEqual(expectedResult);
    });
  });
});
