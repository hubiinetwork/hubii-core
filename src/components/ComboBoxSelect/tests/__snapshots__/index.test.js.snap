// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ComboBoxSelect /> should correctly render 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <ComboBoxSelect
    addInputValidator={[Function]}
    disabled={false}
    handleSelect={[Function]}
    invalidAdditionMessage="message"
    options={
        Array [
            Object {
              "name": "mike",
              "value": "0x324234",
            },
            Object {
              "name": "jake",
              "value": "0x444224",
            },
          ]
    }
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "children": Array [
        <Option
          value="0x324234"
>
          mike
</Option>,
        <Option
          value="0x444224"
>
          jake
</Option>,
      ],
      "disabled": false,
      "filterOption": [Function],
      "notFoundContent": null,
      "onBlur": [Function],
      "onInputKeyDown": [Function],
      "onSearch": [Function],
      "onSelect": [Function],
      "recipient": true,
      "showSearch": true,
      "value": "",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": "mike",
        "nodeType": "class",
        "props": Object {
          "children": Array [
            undefined,
            "mike",
          ],
          "value": "0x324234",
        },
        "ref": null,
        "rendered": Array [
          undefined,
          "mike",
        ],
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": "jake",
        "nodeType": "class",
        "props": Object {
          "children": Array [
            undefined,
            "jake",
          ],
          "value": "0x444224",
        },
        "ref": null,
        "rendered": Array [
          undefined,
          "jake",
        ],
        "type": [Function],
      },
    ],
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "children": Array [
          <Option
            value="0x324234"
>
            mike
</Option>,
          <Option
            value="0x444224"
>
            jake
</Option>,
        ],
        "disabled": false,
        "filterOption": [Function],
        "notFoundContent": null,
        "onBlur": [Function],
        "onInputKeyDown": [Function],
        "onSearch": [Function],
        "onSelect": [Function],
        "recipient": true,
        "showSearch": true,
        "value": "",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": "mike",
          "nodeType": "class",
          "props": Object {
            "children": Array [
              undefined,
              "mike",
            ],
            "value": "0x324234",
          },
          "ref": null,
          "rendered": Array [
            undefined,
            "mike",
          ],
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": "jake",
          "nodeType": "class",
          "props": Object {
            "children": Array [
              undefined,
              "jake",
            ],
            "value": "0x444224",
          },
          "ref": null,
          "rendered": Array [
            undefined,
            "jake",
          ],
          "type": [Function],
        },
      ],
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
